File: src/foam/util/concurrent/AbstractAssembly.java
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

// TODO?: Could provide a reset() method if wanted to make reusable to avoid GC.

/**
 * Abstract implementation of Assembly interface.
 * Provides concurrency support.
 **/
public abstract class AbstractAssembly
  implements Assembly
{
  /** True once endJob() has executed. **/
  boolean complete_ = false;

  /** True if no newe Assembly is waiting for this task to complete. **/
  boolean isLast_   = true;

  public void startJob() {
   // Template method, override in subclass if desired
  }

  public void executeJob() {
   // Template method, override in subclass if desired
  }

  public void endJob() {
   // Template method, override in subclass if desired
  }

  public synchronized boolean isLast() {
   return isLast_;
  }

  public synchronized void complete() {
   complete_ = true;
   notify();
  }

  public synchronized void waitToComplete() {
   isLast_ = false;
   while ( ! complete_ ) {
     try {
       wait();
     } catch (InterruptedException e) {
       // NOP
     }
   }
  }
}

File: src/foam/util/concurrent/AbstractAssembly.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.util.concurrent',
  name: 'AbstractAssembly',
  //abstract: true,
  // TODO?: Could provide a reset() method if wanted to make reusable to avoid GC.

  /**
   * Abstract implementation of Assembly interface.
   * Provides concurrency support.
   **/

  implements: [
    'foam.util.concurrent.Assembly'
  ],

  properties: [
    {
      name: 'isCompleted',
      class: 'Boolean',
      synchronized: true,
      value: false,
      documentation: 'True once endJob() has executed.'
    }
  ],

  methods: [
    {
      name: 'requestLocks',
      javaCode: `
        // Template method, override in subclass if desired
        return null;
      `
    },
    {
      name: 'executeUnderLock',
      javaCode: `
        // Template method, override in subclass if desired
      `
    },
    {
      name: 'startJob',
      javaCode: `
        // Template method, override in subclass if desired
      `
    },
    {
      name: 'executeJob',
      javaCode: `
        // Template method, override in subclass if desired
      `
    },
    {
      name: 'endJob',
      args: [ 'boolean isLast' ],
      javaCode: `
        // Template method, override in subclass if desired
      `
    },
    {
      name: 'complete',
      synchronized: true,
      javaCode: `
         setIsCompleted(true);
         notify();
      `
    },
    {
      name: 'waitToComplete',
      synchronized: true,
      javaCode: `
         while ( ! getIsCompleted() ) {
           try {
             wait();
           } catch (InterruptedException e) {
             // NOP
             Thread.currentThread().interrupt();
           }
         }
      `
    }
  ]
});
